File: adapters/github_action/annotations.py
"""GitHub Actions workflow command emitter.

The runner parses lines like

    ::error file=path,line=N,col=M::message

on stdout and turns them into PR file annotations. See:
https://docs.github.com/actions/using-workflows/workflow-commands-for-github-actions
"""
from __future__ import annotations

import sys
from typing import IO


def _escape_message(s: str) -> str:
    # GitHub's workflow command parser eats `%`, `\r`, `\n`. Escape them
    # so multi-line analyzer reasons survive the round trip.
    return s.replace("%", "%25").replace("\r", "%0D").replace("\n", "%0A")


def _escape_property(s: str) -> str:
    # Property values additionally need `,` and `:` escaped.
    return _escape_message(s).replace(",", "%2C").replace(":", "%3A")


def _emit(
    level: str,
    message: str,
    file: str | None = None,
    line: int | None = None,
    col: int | None = None,
    title: str | None = None,
    out: IO[str] | None = None,
) -> None:
    out = out or sys.stdout
    props: list[str] = []
    if file is not None:
        props.append(f"file={_escape_property(file)}")
    if line is not None:
        props.append(f"line={int(line)}")
    if col is not None:
        props.append(f"col={int(col)}")
    if title is not None:
        props.append(f"title={_escape_property(title)}")
    head = f"::{level}"
    if props:
        head += " " + ",".join(props)
    out.write(f"{head}::{_escape_message(message)}\n")


def error(message: str, **kw) -> None:
    _emit("error", message, **kw)


def warning(message: str, **kw) -> None:
    _emit("warning", message, **kw)


def notice(message: str, **kw) -> None:
    _emit("notice", message, **kw)

File: internal/ghaction/annotations.go
// Package ghaction emits GitHub Actions workflow commands.
package ghaction

import (
	"fmt"
	"io"
	"os"
	"strings"
)

// escapeMessage escapes characters the workflow-command parser eats.
func escapeMessage(s string) string {
	s = strings.ReplaceAll(s, "%", "%25")
	s = strings.ReplaceAll(s, "\r", "%0D")
	s = strings.ReplaceAll(s, "\n", "%0A")
	return s
}

// escapeProperty escapes property values (additionally `,` and `:`).
func escapeProperty(s string) string {
	s = escapeMessage(s)
	s = strings.ReplaceAll(s, ",", "%2C")
	s = strings.ReplaceAll(s, ":", "%3A")
	return s
}

// Annotation captures the optional fields of a workflow command.
type Annotation struct {
	File  string
	Line  int
	Col   int
	Title string
	Out   io.Writer
}

func emit(level, message string, a Annotation) {
	out := a.Out
	if out == nil {
		out = os.Stdout
	}
	var props []string
	if a.File != "" {
		props = append(props, "file="+escapeProperty(a.File))
	}
	if a.Line > 0 {
		props = append(props, fmt.Sprintf("line=%d", a.Line))
	}
	if a.Col > 0 {
		props = append(props, fmt.Sprintf("col=%d", a.Col))
	}
	if a.Title != "" {
		props = append(props, "title="+escapeProperty(a.Title))
	}
	head := "::" + level
	if len(props) > 0 {
		head += " " + strings.Join(props, ",")
	}
	fmt.Fprintf(out, "%s::%s\n", head, escapeMessage(message))
}

func Error(message string, a Annotation)   { emit("error", message, a) }
func Warning(message string, a Annotation) { emit("warning", message, a) }
func Notice(message string, a Annotation)  { emit("notice", message, a) }
